### PR TITLE
Resolve companion from TypeRepr

### PR DESCRIPTION
**Pull request by cchantep**
Implements https://github.com/lampepfl/dotty-feature-requests/issues/244

### CODE REVIEW

File: tests/run-macros/i13798/Macros_1.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+inline def testCompanion: Unit = ${ testCompanionImpl }
+
+def testCompanionImpl(using Quotes): Expr[Unit] = {
+  import quotes.reflect.*
+
+  val tpe = TypeRepr.of[Tuple]
+  val sym: TypeRepr = tpe.companion
+  val repr = sym.typeSymbol.fullName
```
[the remark this answers is not in the document]
Comment (cchantep) [reply tests/run-macros/i13798/Macros_1.scala:10]: Up

[the remark this answers is not in the document]
Comment (cchantep) [reply tests/run-macros/i13798/Macros_1.scala:10]: up

[the remark this answers is not in the document]
Comment (nicolasstucki) [reply tests/run-macros/i13798/Macros_1.scala:10]: Companions are associated with symbol, not types. There are may cases were it is not clear what the compassion of a type would be. Consider the following examples
```scala
TypeRepr.of[Any].companion
TypeRepr.of[AnyRef].companion
TypeRepr.of[AnyVal].companion
TypeRepr.of[AnyKind].companion
TypeRepr.of[x.type].companion
TypeRepr.of[1].companion
TypeRepr.of["abc"].companion
TypeRepr.of[Tuple | Option[Int]].companion
TypeRepr.of[Tuple & Option[Int]].companion
TypeRepr.of[Null].companion
TypeRepr.of[Nothing].companion
```
Not sure what any of those should return.

On the symbol level it is clear. if the symbol does not have a companion we can return `noSymbol`.

**Comment by ckipp01** [comment]
Hey @cchantep / @nicolasstucki what's the current status on this? Is this something you still plan on returning to?

**Comment by nicolasstucki** [comment]
We already support this (see https://github.com/lampepfl/dotty/pull/13801#discussion_r735356999). This is not an operation that makes sense on types. It should be performed on symbols.

**Comment by cchantep** [comment]
> "This is not an operation that makes sense on types"

What make you thinking that?

Once again "the need to resolve the TypeRepr for the symbol, not the companion Symbol"

**Pull request closed without merging**